File: 202.cpp
class Solution {
public:
    int next_num(int n){
        int next = 0;
        int inter; 
        while(n!= 0){
            inter = n%10; 
            next += (inter*inter); 
            n /=10; 
        }
        return next; 
    }
    //applying floyd's cycle
    bool isHappy(int n) {
        int slow = n;
        int fast = next_num(n); 
        while(true){
            cout << "fast "<<fast << endl; 
            cout << "slow "<<slow << endl; 
            if (fast == 1 || slow == 1) return true; 
            if (slow == fast) return false;   
            fast = next_num(next_num(fast));
            slow = next_num(slow); 
        }
    }
};

File: 202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        currNumber = str(n)
        prevNumber = -1
        s = set([])
        while prevNumber != currNumber: 
            prevNumber = currNumber
            currNumber = 0
            for i,l in enumerate(prevNumber): 
                currNumber += int(l)*int(l)
            
            if currNumber == 1:
                return 1 
            if currNumber in s:
                return False 
            else: 
                s.add(currNumber)
            
            currNumber = str(currNumber)
            
        return False 
            
